### ISSUE TITLE
Why long audio get poor performance?

### ISSUE THREAD

**Issue by zjuturtle** [facebookresearch/flashlight#265]
### Question
I found that decode long audio is much worse than split long audio into multi parts. I use librispeech as dataset and test with `decode_transformer_s2s_ngram.cfg`. e.g.

`LibriSpeech/test-clean/1995/1836/1995-1836-0004.flac` is an audio with 33.91 seconds, its groundTruth text 

```
as she awaited her guests she surveyed the table with both satisfaction and disquietude for her social functions were few tonight there were she checked them off on her fingers sir james creighton the rich english manufacturer and lady creighton mister and missus vanderpool mister harry cresswell and his sister john taylor and his sister and mister charles smith whom the evening papers mentioned as likely to be united states senator from new jersey a selection of guests that had been determined unknown to the hostess by the meeting of cotton interests earlier in the day
```

The decode result is

```
as she awaited her guest she surveyed the table with both satisfaction and disquietude for her social functions were few to night there were she checked them off on her fingers sir james clinton the rich english manufacturer and lady horton mister and missus vanderpole mister harry cresswell and his sister john taylor and his sister and mister charles smith whom the evening papers mentioned as likely to be united states senator from new jersey a selection of guests that had been determined unknown to the hostess and mister charles smith whom the evening papers mentioned as likely to be united states senator from
```

Note the ending part `evening papers mentioned as likely to be united states senator from` is far from ground truth `determined unknown to the hostess by the meeting of cotton interests earlier in the day`

But if I break this long audio into 2 parts as following

```
// first part ground truth
as she awaited her guests she surveyed the table with both satisfaction and disquietude for her social functions were few tonight there were she checked them off on her fingers sir james creighton the rich english manufacturer and lady creighton mister and missus vanderpool
```

```
//latter part ground truth
mister harry cresswell and his sister john taylor and his sister and mister charles smith whom the evening papers mentioned as likely to be united states senator from new jersey a selection of guests that had been determined unknown to the hostess by the meeting of cotton interests earlier in the day
```

The decode result is better as following

```
//first part decode
as she awaited her guest she surveyed the table with both satisfaction and disquietude for her social functions were few to night there were she checked them off on her fingers sir james clinton the rich english manufacturer and lady horton mister and missus vanderpole
```

```
//latter part decode
mister harry creswell and his sister john taylor and his sister and mister charles smith whom the evening papers mentioned as likely to be united states senator from new jersey a selection of guests that had been determined unknown to the hostess by the meeting of cotton interests earlier in the day
```

For my result, 

|                    | WER   | LER   |
|--------------------|-------|-------|
| Decode whole audio | 22.9% | 15.3% |
| Decode first part  | 15.1% | 8.9%  |
| Decode latter part | 15.5% | 10.1  |

Is this behavior OK for decoding?

**Comment by zjuturtle** [comment]
Anyone has same issue?

**Comment by tlikhomanenko** [comment]
@zjuturtle we have observed this issue for transformer, investigating it, but will have results in several weeks probably only. Will update here as have any results on debugging.

**Comment by xiaosdawn** [comment]
Hi, I had tested transformer_s2s trained by myself. And I found some duplicate error results:
> sample: test-clean-4507-16021-0026
|T|: to keep afloat and to rescue from oblivion to hold above the gulf were it but a fragment of some language which man has spoken and which would otherwise be lost that is to say one of the elements good or bad of which civilization is composed or by which it is complicated to extend the records of social observation is to serve civilization itself
|P|: to keep afloat and to rescue from oblivion to hold above the gulf were it but a fragment of some language which man has spoken and which would otherwise be lost that is to say one of the elements good or bad of which civilization is composed or by which it is complicated to extend the records of social observation is to serve civilisation itself hold above the gulf were it but a fragment of the gulf were it but a fragmentary to extend the records of social observation is composed or by which it is complicated to extend the records of social observation is composed or by which it is

I used `flashlight commit 8a8f9a2b1` to train the `transformer_s2s` model. The `train.cfg` details are:
`--runname=am_transformer_seq2seq_librispeech
--rundir=/app/w2l-share/AM_rundir
--archdir=/app/w2l-share/configs
--arch=am_arch/am_transformer_s2s.arch
--tokensdir=/app/w2l-share/Download
--tokens=librispeech-train-all-unigram-10000.tokens
--lexicon=/app/w2l-share/Download/librispeech-train+dev-unigram-10000-nbest10.lexicon
--train=/app/w2l-share/lists/train-clean-100.lst,/app/w2l-share/lists/train-clean-360.lst,/app/w2l-share/lists/train-other-500.lst
--valid=dev-clean:/app/w2l-share/lists/dev-clean.lst,dev-other:/app/w2l-share/lists/dev-other.lst
--criterion=transformer
--mfsc
--usewordpiece=true
--wordseparator=_
--am_decoder_tr_dropout=0.1
--am_decoder_tr_layerdrop=0.1
--am_decoder_tr_layers=6
--maxdecoderoutputlen=120
--labelsmooth=0.05
--dataorder=output_spiral
--inputbinsize=25
--attnWindow=softPretrain
--softwstd=4
--trainWithWindow=true
--pretrainWindow=2000
--attention=keyvalue
--encoderdim=256
--memstepsize=5000000
--eostoken=true
--pcttraineval=1
--pctteacherforcing=99
--sampletarget=0.01
--netoptim=adagrad
--critoptim=adagrad
--lr=0.03
--lrcrit=0.03
--linseg=0
--momentum=0.0
--maxgradnorm=0.1
--onorm=target
--sqnorm
--nthread=6
--batchsize=4
--filterbanks=80
--enable_distributed
--warmup=1000000
--saug_start_update=1000000
--lr_decay=200
--lr_decay_step=40
`
Then, using `Decode` for decoding, the 'decode_transformer_s2s_gcnn_clean.cfg' details are:
`--am=/app/w2l-share/librispeech_am_rundir/am_transformer_seq2seq_librispeech/005_model_dev-clean.bin
--tokensdir=/app/w2l-share/Download
--tokens=librispeech-train-all-unigram-10000.tokens
--lexicon=/app/w2l-share/Download/decoder-unigram-10000-nbest10.lexicon
--lm=/app/w2l-share/Download/lm_librispeech_wp_10k_gcnn_14B.bin
--datadir=/app/w2l-share/lists
--test=test-clean.lst
--uselexicon=false
--sclite=/app/w2l-share/librispeech_am_rundir/am_transformer_seq2seq_Librispeech/test_005_model_dev-clean.wpGcnn
--decodertype=tkn
--lmtype=convlm
--lm_vocab=/app/w2l-share/Download/lm_librispeech_wp_10k_gcnn_14B.vocab
--lm_memory=5000
--beamsize=50
--beamsizetoken=10
--beamthreshold=10
--nthread_decoder=1
--show
--showletters
--lmweight=0.2859322510812
--eosscore=-6.4242331182886
` 
BTW, the gcnn LM is downloaded through the link [wp GCNN](https://dl.fbaipublicfiles.com/wav2letter/sota/2019/lm/lm_librispeech_wp_10k_gcnn_14B.bin). 

In addtion, before using `flashlight commit 8a8f9a2b1` to train AM, I had trained `transformer_ctc` using `wav2letter v0.2 commit 6d7d3c19`. But the error mentioned above did not occur. 
>sample: test-clean-4507-16021-0026
|T|: to keep afloat and to rescue from oblivion to hold above the gulf were it but a fragment of some language which man has spoken and which would otherwise be lost that is to say one of the elements good or bad of which civilization is composed or by which it is complicated to extend the records of social observation is to serve civilization itself
|P|: to keep afloat and to rescue from oblivion to hold above the gulf were it but a fragment of some language which man has spoken and which would otherwise be lost that is to say one of the elements good or bad of which civilization is composed or by which it is complicated to extend the records of social observation is to serve civilization itself

So, I don't know if the error is the same as what @zjuturtle said.
Any suggestions would be appreciated.

**Comment by tlikhomanenko** [comment]
Nope, this is unrelated thing @xiaosdawn. The problem is in s2s. CTC model has no problems with repetitions because of loss nature, while s2s with attention generates output based also on the inferred prefix. This is known problem of s2s models: for some sentences it is solved, for some not. Eosscore in our decoder is trying to solve this issue, but still some sentences could have this problem.

**Comment by xiaosdawn** [comment]
Got it. Thanks.

**Comment by xiaosdawn** [comment]
Recently, I have tested Convnet with ctc trained using librispeech datasets. And found some words duplicate errors, eg:
``` |T|: at that epoch of pristine simplicity however matters of even slighter public interest and of far less intrinsic weight than the welfare of hester and her child were strangely mixed up with the deliberations of legislators and acts of state
  |P|: at that epoch epoch epoch of pristine simplicity simplicity simplicity however however matters of even slighter public public interest and of far less less intrinsic weight than the welfare of hester and and her child were strangely mixed up with the deliberations of legislators and acts of state
  |t|: a t _ t h a t _ e p o c h _ o f _ p r i s t i n e _ s i m p l i c i t y _ h o w e v e r _ m a t t e r s _ o f _ e v e n _ s l i g h t e r _ p u b l i c _ i n t e r e s t _ a n d _ o f _ f a r _ l e s s _ i n t r i n s i c _ w e i g h t _ t h a n _ t h e _ w e l f a r e _ o f _ h e s t e r _ a n d _ h e r _ c h i l d _ w e r e _ s t r a n g e l y _ m i x e d _ u p _ w i t h _ t h e _ d e l i b e r a t i o n s _ o f _ l e g i s l a t o r s _ a n d _ a c t s _ o f _ s t a t e
  |p|: _ a t _ t h a t _ e p o c h _ o f _ p r i s t i n e _ s i m p l i c i t y _ h o w e v e r _ m a t t e r s _ o f _ e v e n _ s l i g h t e r _ p u b l i c _ i n t e r e s t _ a n d _ o f _ f a r _ l e s s _ i n t r i n s i c _ w e i g h t _ t h a n _ t h e _ w e l f a r e _ o f _ h e s t e r _ a n d _ h e r _ c h i l d _ w e r e _ s t r a n g e l y _ m i x e d _ u p _ w i t h _ t h e _ d e l i b e r a t i o n s _ o f _ l e g i s l a t o r s _ a n d _ a c t s _ o f _ s t a t e
  [sample: test-clean-1221-135767-0002, WER: 20%, LER: 0.41841%, slice WER: 20%, slice LER: 0.41841%, decoded samples (thread 4): 1]
```
Then, I tried different parameters for decoding, and got some results as:
| duplicate  | decodertype | uselexicon | LexiconDecoder | LexiconFreeDecoder |
| ------------ | ----------- | ---------- | -------------- | ------------------ |
| no            | wrd         | false      | √              |                    |
| yes           | wrd         | true       | √              |                    |
| no           | tkn         | false      |                | √                  |
| yes           | tkn         | true       | √              |                    |

When `no duplicate errors`, the results are all right.

So, I was more confused for the duplicate errors. 
Are there any further solutions or suggestions to this problems?

**Comment by xiaosdawn** [comment]
Anyone has same issue or some suggestions?

**Comment by tlikhomanenko** [comment]
Could you provide flashlight commit and the model with cmd line to repro this deplications? There could be a bug in postprocessing.  

I will have a look anyway in upcoming days, will update as will have a repro and fix.

**Comment by xiaosdawn** [comment]
Thank @tlikhomanenko . 
The flashlight commit is [36b242e](https://github.com/facebookresearch/flashlight/commit/36b242e476ff0a80468751d30c6c550cb9ed9877).
The model was trained by [streaming_convets](https://github.com/facebookresearch/wav2letter/tree/master/recipes/streaming_convnets/librispeech) recipe with `librispeech`
``` ./asr_train train --flagsfile ../train_am_convnets_500ms_future_context.cfg"

The `train_am_convnets_500ms_future_context.cfg` details as:
``` --runname=am_convnet_inference_2019
--rundir=/app/w2l-share/AM_rundir
--archdir=/app/w2l-share/configs
--arch=am_arch/am_500ms_future_context.arch
--tokensdir=/app/w2l-share/Lexicon-and-Tokens/Download
--tokens=librispeech-train-all-unigram-10000.tokens
--lexicon=/app/w2l-share/Lexicon-and-Tokens/Download/librispeech-train+dev-unigram-10000-nbest10.lexicon
--train=/app/w2l-share/lists/lsts-processed/train-clean-100.lst,/app/w2l-share/lists/lsts-processed/train-clean-360.lst,/app/w2l-share/lists/lsts-processed/train-other-500.lst
--valid=dev-clean:/app/w2l-share/lists/lsts-processed/dev-clean.lst,dev-other:/app/w2l-share/lists/lsts-processed/dev-other.lst
--criterion=ctc
--batchsize=8
--lr=0.4
--momentum=0.0
--maxgradnorm=0.5
--reportiters=1000
--nthread=6
--mfsc=true
--usewordpiece=true
--wordseparator=_
--filterbanks=80
--minisz=200
--mintsz=2
--maxisz=33000
--enable_distributed=true
--pcttraineval=1
--minloglevel=0
--logtostderr
--onorm=target
--sqnorm
--localnrmlleftctx=300
--lr_decay=10000


Then decoded the `convnet` model with:
``` ./asr_decode --flagsfile=../decode_500ms_right_future_ngram_other.cfg

`decode_500ms_right_future_ngram_other_all.cfg` details as:
``` --am=/app/w2l-share/AM_rundir/am_convnet_inference_2019/003_model_dev-other.bin
--tokensdir=/app/w2l-share/Lexicon-and-Tokens/Download
--tokens=librispeech-train-all-unigram-10000.tokens
--lexicon=/app/w2l-share/Lexicon-and-Tokens/Download/decoder-unigram-10000-nbest10.lexicon4test
--datadir=/app/w2l-share/lists/lsts-processed
--test=test-other.lst
--uselexicon=true
--sclite=/app/w2l-share/AM_rundir/am_convnet_inference_2019/test_003_model_dev-other
--decodertype=wrd
--lmtype=kenlm
--silscore=0
--beamsize=500
--beamsizetoken=100
--beamthreshold=100
--nthread_decoder=8
--smearing=max
--show
--showletters

**Comment by tlikhomanenko** [comment]
Seems you are using very old commit, could you try to build recent and check if it still gives you duplication words?

**Comment by xiaosdawn** [comment]
OK. I'll use the lastest commit. 
But I don't have too much machines and it's too slow to train. If I have some results, I'll tell you.
Thanks again.

**Comment by tlikhomanenko** [comment]
You can rebuild and just run decoding, no need to retrain.

**Comment by xiaosdawn** [comment]
Thank you very much.
I rebuild flashlight using Dockerfile. Then run `build/bin/asr/fl_asr_decode` . But I sitll got some duplication words:
```
|T|: at that epoch of pristine simplicity however matters of even slighter public interest and of far less intrinsic weight than the welfare of hester and her child were strangely mixed up with the deliberations of legislators and acts of state
|P|: at that epoch epoch of pristine simplicity simplicity simplicity however however matters of even slighter public interest and of far less less intrinsic weight than the welfare of hester and and her child were strangely mixed up with the deliberations of legislators and acts of state
|t|: a t _ t h a t _ e p o c h _ o f _ p r i s t i n e _ s i m p l i c i t y _ h o w e v e r _ m a t t e r s _ o f _ e v e n _ s l i g h t e r _ p u b l i c _ i n t e r e s t _ a n d _ o f _ f a r _ l e s s _ i n t r i n s i c _ w e i g h t _ t h a n _ t h e _ w e l f a r e _ o f _ h e s t e r _ a n d _ h e r _ c h i l d _ w e r e _ s t r a n g e l y _ m i x e d _ u p _ w i t h _ t h e _ d e l i b e r a t i o n s _ o f _ l e g i s l a t o r s _ a n d _ a c t s _ o f _ s t a t e
|p|: _ a t _ t h a t _ e p o c h _ o f _ p r i s t i n e _ s i m p l i c i t y _ h o w e v e r _ m a t t e r s _ o f _ e v e n _ s l i g h t e r _ p u b l i c _ i n t e r e s t _ a n d _ o f _ f a r _ l e s s _ i n t r i n s i c _ w e i g h t _ t h a n _ t h e _ w e l f a r e _ o f _ h e s t e r _ a n d _ h e r _ c h i l d _ w e r e _ s t r a n g e l y _ m i x e d _ u p _ w i t h _ t h e _ d e l i b e r a t i o n s _ o f _ l e g i s l a t o r s _ a n d _ a c t s _ o f _ s t a t e
[sample: test-clean-1221-135767-0002, WER: 15%, TER: 0.41841%, slice WER: 10.4425%, slice TER: 2.59862%, decoded samples (thread 0): 190]
```
the `decode.cfg` details as:
```
--am=/app/w2l-share/AM_rundir/am_convnet_inference_2019/004_model_dev-clean.bin
--tokens=/app/w2l-share/Lexicon-and-Tokens/Download/librispeech-train-all-unigram-10000.tokens
--lexicon=/app/w2l-share/Lexicon-and-Tokens/Download/decoder-unigram-10000-nbest10.lexicon4test
--datadir=/app/w2l-share/lists/lsts-processed
--test=test-clean.lst
--uselexicon=true
--sclite=/app/w2l-share/AM_rundir/am_convnet_inference_2019/test_004_model_dev-clean_useLtrue_wrd
--decodertype=wrd
--lmtype=kenlm
--silscore=0
--beamsize=500
--beamsizetoken=100
--beamthreshold=100
--nthread_decoder=1
--smearing=max
--show
--showletters
```

**Comment by tlikhomanenko** [comment]
@xiaosdawn could you share the model? then I can repro this and debug the error. Also could you specify what lm are you using (I don't see --lm flag in your config)?

**Comment by xiaosdawn** [comment]
My trained `convnet` can be accessed [convnet_model_dev-clean.bin](https://drive.google.com/file/d/1V6UsG7fhy7wryS0-KGb9Gh_oyd1ruFt0/view?usp=sharing)
The `decode.cfg` is the same as [decode_500ms_right_future_ngram_other.cfg](https://github.com/facebookresearch/wav2letter/blob/master/recipes/streaming_convnets/librispeech/decode_500ms_right_future_ngram_other.cfg). So the `lm` is default, `zero`.

**Comment by tlikhomanenko** [comment]
Thanks, will try to debug this week and find the issue, will update asap have something.

**Comment by tlikhomanenko** [comment]
Ok, finally we found the issue with @xuqiantong and @vineelpratap.

Here is the fix which we gonna to land, but you can use it right now to continue your work. Please add this lines after https://github.com/facebookresearch/flashlight/blob/master/flashlight/lib/text/decoder/LexiconDecoder.cpp#L113
```
          if (prevLex == lexicon_->getRoot() && prevHyp.token == n) {
            // if previous token == word, don't repeat token
            continue;
          }

```

**Comment by xiaosdawn** [comment]
It works. This fixes the issue for me. 
Thanks again for your team's efforts @tlikhomanenko

**Comment by tlikhomanenko** [comment]
Cool! Also this fix is landed, so feel free to switch to the latest master. Closing the issue.
